src/dune_rules/pkg_rules.ml

```diff
@@ -1798,7 +1809,46 @@ let build_rule context_name ~source_deps (pkg : Pkg.t) =
       |> Action_builder.return
       |> Action_builder.with_no_targets
     in
-    Action_builder.progn (build_and_install @ [ install_file_action ])
+    (* This action will print a "Downloading" message for the package,
+       only if its source directory is not yet present. Not all
+       packages have source archives and the "Downloading" message that
+       prints while fetching the source arcive won't have been printed
+       for these packages. This is to avoid the situation where dune
```

Comment:
You're saying that this situation is confusing, but I think it reflects exactly what dune is doing and the attempt to workaround as confusing. It's indeed the case that dune will have to rebuild a package without re-downloading the sources. Why should we pretend otherwise?

Reply:
I should add that there's a far more common case for rebuilding a package without re-dowloading its sources. Consider upgrade a dependency of package. You will need download the dependency, build it, and then download the package itself without re-downloading.

> Not all packages have source archives and the "Downloading" message that prints while fetching the source arcive won't have been printed for these packages

A better way to solve this would be to modify the fetch rules to insert a "download" action before fetching.

Reply:
For the sake of simplicity/accuracy I've removed this message for now. Now dune will only print a "downloading" message when a package's source archive is downloaded. We'll be testing this feature in some user interviews so if it turns out to confuse people we can find a way to solve that.

Regarding extra sources, for the initial version of this feature I don't think it's necessary to print anything. I'm hesitant to add a "download" message before fetching as we do with the source archive as it's common for packages to have multiple extra sources and I don't want to print the message for each one. There's a bunch of ways we could solve that problem but I want to get the basic feature merged before thinking about that.

Reply:
It will be even simpler and more correct if you remove the `output` parameter. I don't really understand this explanation:

> The [output] argument is the path of a file that will be created by the operation that the progress message is about so that the message can be elided if the file already exists.

Why do you want to not print the message if the target directory already exists? This is not the criteria for rebuilding packages. Essentially, you're omitting this progress message whenever the package is rebuilt after the first time.

Reply:
In a previous implementation of this feature I found that it would print the progress messages for each dependency, even if they didn't get (re)built/downloaded, and the `output` param was a workaround to prevent this. However I just checked and now that printing status messages is implemented as an action this doesn't happen, so there's no reason to keep the `output` parameter. I've removed it.